### PR TITLE
Toggle favorite reports/dashboards from table view

### PR DESCRIPTION
**Pull request by 100starnight**
## Screenshots
pr1531_newerScreenshot

## License
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/directory/addon/components/dir-item-favorite-cell.hbs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{{!-- Copyright 2020, Yahoo Holdings Inc. Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms. --}}
```
Comment (cythrawll) [inline packages/directory/addon/components/dir-item-favorite-cell.hbs:1]: can we just be like @icon={{if @value.isFavorite "star-solid" "star"}}

File: packages/directory/addon/components/dir-item-favorite-cell.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright 2021, Yahoo Holdings Inc.
+ * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
+ *
+ * This component won't be used directly. It is passed to ember-light-table as a custom cell component.
+ * Ember-light-table will pass any parameters in through the value attribute.
+ *
+ * <DirItemNameCell
+ *  @value={{@item}}
+ * />
```
Comment (kevinhinterlong) [inline packages/directory/addon/components/dir-item-favorite-cell.ts:10]:
```suggestion
```

File: packages/directory/addon/components/dir-item-favorite-cell.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ */
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import type NaviNotificationsService from 'navi-core/services/interfaces/navi-notifications';
+import type UserService from 'navi-core/services/user';
+import ReportModel from 'navi-core/models/report';
+import DashboardModel from 'navi-core/models/dashboard';
```
Comment (kevinhinterlong) [inline packages/directory/addon/components/dir-item-favorite-cell.ts:17]:
```suggestion
import type ReportModel from 'navi-core/models/report';
import type DashboardModel from 'navi-core/models/dashboard';
```

File: packages/directory/addon/components/dir-item-name-cell.ts [hunk 4]
```diff
@@ -10,11 +10,12 @@
  * />
  */
 import Component from '@glimmer/component';
-import FileTypes from 'navi-directory/utils/enums/file-types';
 import { pluralize } from 'ember-inflector';
+import ReportModel from 'navi-core/models/report';
+import DashboardModel from 'navi-core/models/dashboard';
```
Comment (kevinhinterlong) [inline packages/directory/addon/components/dir-item-name-cell.ts:15]:
```suggestion
import type ReportModel from 'navi-core/models/report';
import type DashboardModel from 'navi-core/models/dashboard';
```

File: packages/directory/addon/components/dir-table.ts [hunk 5]
```diff
@@ -72,7 +76,7 @@ export default class DirTableComponent extends Component<DirTableComponentArgs>
         classNames: 'dir-table__header-cell dir-table__header-cell--actions',
         cellComponent: 'dir-asset-row-actions',
         cellClassNames: 'dir-table__cell dir-table__cell--actions',
-        breakpoints: ['tablet', 'desktop', 'jumbo'],
+        breakpoints: ['mobile', 'tablet', 'desktop', 'jumbo'],
```
Comment (kevinhinterlong) [inline packages/directory/addon/components/dir-table.ts:79]: Does this look ok on mobile?

File: packages/directory/tests/acceptance/dir-sidebar-toggle-test.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import { findAll, click, currentURL, visit } from '@ember/test-helpers';
 import { module, test } from 'qunit';
 import { setupApplicationTest } from 'ember-qunit';
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
```
Comment (kevinhinterlong) [inline packages/directory/tests/acceptance/dir-sidebar-toggle-test.ts:4]: We could add an override in eslint to allow this in test files
https://github.com/yavin-dev/framework/blob/03be283da2f7f456af91661ebcde538749ae206b/packages/notifications/.eslintrc.js#L69

File: packages/directory/tests/acceptance/dir-table-test.js [hunk 7]
```diff
@@ -112,4 +112,52 @@ module('Acceptance | dir table', function (hooks) {
...
+    // make sure UI un-favorite is reflected both in stored data & visually
+    assert.ok(favIcon.firstElementChild.classList.contains('d-star'), 'non-favorite item shows a hollow star');
+    favoriteReports = user.favoriteReports.toArray().map((ele) => ele.title);
+    assert.equal(favoriteReports.length, 0, 'user no longer stores item as favorite');
+
+    // re-favorite item
+    await click(favIcon.firstElementChild);
+
+    // make sure UI re-favorite shows up both in stored data & visually
+    assert.ok(favIcon.firstElementChild.classList.contains('d-star-solid'), 'favorite item shows a solid star');
```
Comment (kevinhinterlong) [inline packages/directory/tests/acceptance/dir-table-test.js:140]: We can use `assert.dom(favIcon.firstElementChild).hasClass(...`

File: packages/directory/tests/acceptance/dir-table-test.js [hunk 8]
```diff
@@ -112,4 +112,52 @@ module('Acceptance | dir table', function (hooks) {
       .dom('.dir-table-filter select')
       .hasValue('Dashboards', 'The selected type is set based on the query param in the url');
   });
+
+  test('favorite item toggle', async function (assert) {
+    assert.expect(9);
+    await visit('/directory/my-data');
+    const user = await this.owner.lookup('service:user').findUser();
```
Comment (kevinhinterlong) [inline packages/directory/tests/acceptance/dir-table-test.js:119]: Can we maybe replace the use of the user service with a function that returns a list of all favorite reports from the dom?

```
const getFavorites = () => {
	return findAll('...').filter(el => el.querySelector('.d-star-solid')).map(fav => fav.querySelector('.dir-item-name-cell').textContent.trim())
}

**Review by kevinhinterlong — COMMENTED** [review]
Small stuff

File: packages/directory/addon/components/dir-item-favorite-cell.ts (hunk 3 above)
Comment (100starnight) [reply packages/directory/addon/components/dir-item-favorite-cell.ts:17]: if I import those with type, ts throws an error when I call instanceof

File: packages/directory/addon/components/dir-table.ts (hunk 5 above)
Comment (100starnight) [reply packages/directory/addon/components/dir-table.ts:79]: lol nope! That was supposed to be on the favorite cell, not the action cell. Good catch

**Review by cythrawll — APPROVED** [review]

**Review by kevinhinterlong — APPROVED** [review]

**Pull request merged**